cardinal/ecs/cmdbuffer/cmdbuffer.go

```diff
@@ -0,0 +1,508 @@
+package cmdbuffer
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	"github.com/redis/go-redis/v9"
+	"pkg.world.dev/world-engine/cardinal/ecs/archetype"
+	"pkg.world.dev/world-engine/cardinal/ecs/codec"
+	"pkg.world.dev/world-engine/cardinal/ecs/component"
+	"pkg.world.dev/world-engine/cardinal/ecs/entity"
+	"pkg.world.dev/world-engine/cardinal/ecs/filter"
+	"pkg.world.dev/world-engine/cardinal/ecs/log"
+	"pkg.world.dev/world-engine/cardinal/ecs/storage"
+	"pkg.world.dev/world-engine/cardinal/ecs/store"
+)
+
+var _ store.IManager = &Manager{}
+
+type Manager struct {
+	client *redis.Client
+
+	compValues         map[compKey]any
+	compValuesToDelete map[compKey]bool
+	typeToComponent    map[component.TypeID]component.IComponentType
+
+	activeEntities map[archetype.ID]activeEntities
+
+	// Fields that track the next valid entity ID that can be assigned
+	nextEntityIDSaved uint64
+	pendingEntityIDs  uint64
+	isEntityIDLoaded  bool
+
+	// Archetype ID management.
+	entityIDToArchID       map[entity.ID]archetype.ID
+	entityIDToOriginArchID map[entity.ID]archetype.ID
+
+	archIDToComps  map[archetype.ID][]component.IComponentType
+	pendingArchIDs []archetype.ID
+
+	logger *log.Logger
+}
+
+var (
+	errorArchIDNotFound     = errors.New("archetype for components not found")
+	doesNotExistArchetypeID = archetype.ID(-1)
+)
+
+// NewManager creates a new command buffer manager that is able to queue up a series of states changes and
+// atomically commit them to the underlying redis storage layer.
+func NewManager(client *redis.Client) (*Manager, error) {
+	m := &Manager{
+		client:             client,
+		compValues:         map[compKey]any{},
+		compValuesToDelete: map[compKey]bool{},
+
+		activeEntities: map[archetype.ID]activeEntities{},
+
+		entityIDToArchID:       map[entity.ID]archetype.ID{},
+		entityIDToOriginArchID: map[entity.ID]archetype.ID{},
+
+		// These fields cannot be set until RegisterComponents is called
+		archIDToComps:   nil,
+		typeToComponent: nil,
+	}
+
+	return m, nil
+}
+
+func (m *Manager) RegisterComponents(comps []component.IComponentType) error {
+	m.typeToComponent = map[component.TypeID]component.IComponentType{}
+	for _, comp := range comps {
+		m.typeToComponent[comp.ID()] = comp
+	}
+
+	return m.loadArchIDs()
+}
+
+// AtomicFn commits any pending changes to the DB, performs the actions inside of fn. If fn results in an error,
+// any pending changes that were generated inside of fn are discarded. If fn results in no error, the pending
+// changes are committed.
+func (m *Manager) AtomicFn(fn func() error) error {
```

Comment:
I removed this method entirely. CommitPending and DiscardPending can be used to do the exact same thing. I suspect in practice we are only every really going to use the CommitPending (when ticks successfully complete) and bail out completely instead of calling DiscardPending.